snippets/mark-fields-as-touched.md

```diff
@@ -0,0 +1,36 @@
+---
+title: Mark reactive form fields as touched to highlight validation messages
+level: Intermediate
+tags:
+  - reactive forms validation
+  - tips
+  - good to know
+---
+
+# Content
+Here is the way to notify user that there are fields with non-valid values.
+
+**markFieldsAsTouched** function takes _form_ which is FormGroup or FormArray as arg. 
+
+```typescript
+  function markFieldsAsTouched(form: FormGroup | FormArray): void {
+    Object.values(form.controls).forEach(
```

Comment:
both `FormGroup` and `FormArray` support method `markAsTouched`. Do we really have to iterate through the all controls?

Reply:
Sergey, I'm so sorry for so long response time! I just missed it, sorry

In this function it makes sense to use FormGroup/FormArray markAsTouched as indicator to define that there still are fields with non-valid values when onSubmit was clicked to notify user about validation-errors. 

But I there are cases when developer reuse fields with theirs controls. In this cases developer usually has special component with field (and its controls). So the field doesn't know of the whole FormGroup markAsTouched status, and control has only its own markAsTouched status

If you know what I mean)
I have an article on this subject that can be a good example of encapsulated field-control logic. If you have time, check it out: https://link.medium.com/qejCNDO9VU

If you have any ideas how to use in a good way, please share, I'll be happy to check it out and improve the things that I made before)

Reply:
@alex-bu-93 it took some time to dive into the problem.  `FormGroup.markAsTouched()` doesn't mark children controls
here's the some explanation from developers https://github.com/angular/angular/issues/11774#issuecomment-253396201

I think we need to update snippets with code samples from your article (simplifed) and after it'll be ready to go

Reply:
Sergey,
1) "it took some time to dive into the problem" means that you don't have problems, in case I you have, you'll quickly and easily dive into it
2) "FormGroup.markAsTouched() doesn't mark children controls" means that apparently you haven't understood the function logic. If you do it, you'll see that I reuse the function in "else" block until I get exactly control and mark it as touched
3) "here's the some explanation from developers angular/angular#11774 (comment)". As far as I could understand, this comment have no deal with my PR as I work with controls not group and I solve the problem "How to mark all of them touched", not the problem "What if I need to untouche controls" or "What if I need to untouche form". That could be another one snippet.
4) " think we need to update snippets with code samples from your article (simplifed) and after it'll be ready to go". Again as in my response to you comment on medium, pity that you didn't get the article essence. Article is an approach to work with forms, not the copy-paste code-solution.

Reply:
Once again. `FormGroup` implements `markAsTouched` as part of `AbstractControl`. For reasons described in the comment, they don't mark group controls. This snippet shows how to solve it. Snippets are small solutions for specific problems. You approach for form building is out of the scope this repo and snippet topic.

Reply:
Don't know what you want by your comments. You can suggest code changes here. Use it to be more understandable.